### PR TITLE
Multiple buttons

### PR DESCRIPTION
**Pull request by collnwalkr**
Now able to inject multiple buttons. `locals.action` can either contain an array of objects or a single object. README.md and THEME.md have been updated to reflect this addition.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -205,6 +205,30 @@ var email = {
...
+    body: {
+        action: [
+            {
+                instructions: 'To get started with Mailgen, please click here:',
+                button: {
+                    color: '#22BC66',
+                    text: 'Confirm your account',
+                    link: 'https://mailgen.js/confirm?s=d9729feb74992cc3482b350163a1a010'
+                }
+            },{
```
Comment (eladnava) [inline README.md]: Please have the second object `{` start on a new line.

File: README.md [hunk 2]
```diff
@@ -205,6 +205,30 @@ var email = {
...
+                instructions: 'To read our frequently asked questions, please click here:',
+                button: {
+                    text: 'Read our FAQ',
+                    link: 'https://mailgen.js/faq'
+                }        
+            }],
```
Comment (eladnava) [inline README.md]: Please have the closing `]` on a new line and omit the `,`.

File: THEME.md [hunk 3]
```diff
@@ -205,25 +205,46 @@ It's a good idea to add this to the top of the template to improve the styling o
 
 ## Action Injection
 
-The following will inject the action link (or button) into the e-mail:
+The following will inject the action link (or button) into the e-mail.
 
 ```html
 <!-- Action -->
+<% if(locals.action.constructor !== Array){ %>
```
Comment (eladnava) [inline THEME.md]: Please add a space after `if` and before `(` as per the project's code style. Please also do this in all other places in your PR (in theme files).

File: THEME.md [hunk 4]
```diff
@@ -205,25 +205,46 @@ It's a good idea to add this to the top of the template to improve the styling o
...
+    <% }) -%>
 <% } %>
 ```
 
 It's a good idea to add this to the top of the template to specify a fallback color for the action button in case it wasn't provided by the user:
 
 ```html
 <%
+if (locals.action){
+
```
Comment (eladnava) [inline THEME.md]: Please omit the extra line breaks(s) here and in the theme files, as well as fix the code style (spacing) so it matches the project code style.

File: THEME.md [hunk 5]
```diff
@@ -205,25 +205,46 @@ It's a good idea to add this to the top of the template to improve the styling o
...
+    // If there is only one action, make into array of one actions
+    if(locals.action.constructor !== Array){
+        action = [action];
+    }
+
+    locals.action.forEach(function(actionItem){
+        fallbackColor(actionItem.button);
```
Comment (eladnava) [inline THEME.md]: I think that we should omit the function `fallbackColor` and move its logic to within the `forEach` to simplify the code. What do you think?

File: THEME.md (hunk 3 above)
Comment (eladnava) [inline THEME.md]: Also, the conversion of the `locals.action` to an array should be done at the top of the theme files (I see that you have done that but have included it here as well).

File: themes/cerberus/index.txt [hunk 6]
```diff
@@ -13,8 +13,13 @@
 <% } %>
 
 <% if (locals.action) { %>
-<%- action.instructions %>
-<%- action.button.link %>
+<% if(locals.action.constructor !== Array){ %>
```
Comment (eladnava) [inline themes/cerberus/index.txt]: Please move the conversion of the `locals.action` to an array to the top of the file, and condense it into one line, as follows:
```js
<% if (locals.action && locals.action.constructor !== Array) { action = [action]; } %>
```

We must do this because the plaintext version of the e-mail is sent as is, including any line-breaks added as a result of the `ejs` logic.  The less line breaks we can take up, the less line-bloated the plaintext e-mail will appear.

File: THEME.md [hunk 7]
```diff
@@ -205,25 +205,46 @@ It's a good idea to add this to the top of the template to improve the styling o
 
 ## Action Injection
 
-The following will inject the action link (or button) into the e-mail:
+The following will inject the action link (or button) into the e-mail.
```
Comment (eladnava) [inline THEME.md]: Just wondering, why did you change the grammar to a `.` instead of a `:`?

Comment (collnwalkr) [inline THEME.md]: this was a typo. thanks for catching it!

File: THEME.md (hunk 3 above)
Comment (collnwalkr) [inline THEME.md]: @eladnava I agree that it should be done at the top of the theme files. This was just a safeguard in case someone did not do that. For instance, I'm using someone else's theme that does not have this update but I want multiple buttons. Do you still think this should be removed?

File: THEME.md (hunk 5 above)
Comment (collnwalkr) [inline THEME.md]: I agree. I have fixed this and applied the change across themes.

File: themes/cerberus/index.txt (hunk 6 above)
Comment (collnwalkr) [inline themes/cerberus/index.txt]: Interesting, I didn't know that. When you say top of the file, do you mean line 1 above `<%- title %>` ? That will put a line break above the title.

**Comment by collnwalkr** [comment]
@eladnava made coding styling changes. Let me know if there's anything else!

**Comment by eladnava** [comment]
@collnwalkr I just thought of something -- instead of converting the `action` to an array if it is an object in every theme file, how about we simply implement this once within `index.js`?

Somewhere along these lines:
https://github.com/eladnava/mailgen/blob/461ab1c84ad99488161deb2d71457f2b547e4f56/index.js#L121

See how we already do something similar for `intro` and `outro`?

I'm actually thinking we can change `convertToArrayIfString` to `convertToArray` which will convert any object it receives into an array (if it isn't already) using the logic you implemented in the templates. It shouldn't break anything since `intro` and `outro` are strings anyway, and it's actually safer this way.

You can then use this function to convert `body.action` to an array.

**Comment by collnwalkr** [comment]
@eladnava That's a great idea! I have made those changes and have committed those changes.

File: index.js [hunk 8]
```diff
@@ -117,9 +117,10 @@ Mailgen.prototype.parseParams = function (params) {
         body.title = (body.name ? body.greeting + ' ' + body.name : body.greeting) + ',';
     }
 
-    // Convert intro/outro to arrays if a string is used instead
-    body.intro = convertToArrayIfString(body.intro)
-    body.outro = convertToArrayIfString(body.outro)
+    // Convert intro,outro, and action to arrays if a string or object is used instead
```
Comment (eladnava) [inline index.js]: Missing space between `intro,` and `outro` (a bit OCD, I know) 😄

File: THEME.md [hunk 9]
```diff
@@ -210,20 +210,26 @@ The following will inject the action link (or button) into the e-mail:
...
+    <% action.forEach(function (actionItem) { -%>
+        <p><%- actionItem.instructions %></p>
+        <a href="<%- actionItem.button.link %>" style="color:<%- actionItem.button.color %>" target="_blank">
+            <%- actionItem.button.text %>
+        </a>
+    <% }) -%>
 <% } %>
 ```
 
 It's a good idea to add this to the top of the template to specify a fallback color for the action button in case it wasn't provided by the user:
```
Comment (eladnava) [inline THEME.md]: It would be great if you could update the text here to print `for the action buttons` (plural). 😄

File: THEME.md [hunk 10]
```diff
@@ -209,21 +209,36 @@ The following will inject the action link (or button) into the e-mail:
 
 ```html
 <!-- Action -->
+<% if (locals.action.constructor !== Array) { %>
```
Comment (collnwalkr) [inline THEME.md]: I believe if you look at the latest commit this has already been changed.

Comment (eladnava) [inline THEME.md]: @collnwalkr Oops, my bad, that was an old comment that got submitted for some reason.

File: index.js [hunk 11]
```diff
@@ -29,9 +29,9 @@ function Mailgen(options) {
     this.cacheThemes();
 }
 
-function convertToArrayIfString(data) {
-    // Convert string to single-element array 
-    if (typeof data === 'string') {
+function convertToArray(data) {
+    // Convert object to array
+    if (data.constructor !== Array) {
```
Comment (collnwalkr) [inline index.js]: that's a good idea. fixed.

**Comment by collnwalkr** [comment]
@eladnava those were some great comments! let me know if there's anything else I can change/update.

File: themes/salted/index.html [hunk 12]
```diff
@@ -325,34 +329,36 @@
...
-                                        </table>
-                                    </td>
-                                </tr>
+                            <% action.forEach(function (actionItem) { -%>
+                                    <tr>
+                                        <td>
+                                            <!-- COPY -->
+                                            <table align="center" width="100%" border="0" cellspacing="0" cellpadding="0">
+                                                <tr>
+                                                    <td align="center" style="padding: 0 0 0 0; font-size: 16px; line-height: 25px; font-family: 'Helvetica Neue', Helvetica, Arial, sans-serif; color: #666666;" class="padding-copy">
```
Comment (eladnava) [inline themes/salted/index.html]: A little visual "bug" in the salted theme:

screen shot 2016-10-01 at 11 30 20 pm

There is insufficient padding between the action items.

We should change the `padding` declaration in line 338 and specify `25px` for the top, to avoid this situation.

We should then remove `25px` padding from the parent `<td>` that contains the action buttons so that the padding-top isn't too high, in line 325.

The new padding for the parent `<td>` (L325):
```
padding: 15px 15px 40px 15px;
```

Which gives us:
screen shot 2016-10-01 at 11 35 27 pm

File: themes/neopolitan/index.html [hunk 13]
```diff
@@ -261,37 +265,40 @@
...
+                                    style="background-color:<%- actionItem.button.color %>;border-radius:4px;color:#ffffff;display:inline-block;font-family:Helvetica, Arial, sans-serif;font-size:16px;font-weight:bold;line-height:50px;text-align:center;text-decoration:none;width:200px;-webkit-text-size-adjust:none;">
+                                    <%- actionItem.button.text %>
+                                    </a>
+                                        <!--[if mso]>
+                                        </center>
+                                        </v:roundrect>
+                                    <![endif]--></div>
                                     <br>
-                                    <%- action.instructions %>
                                     <br>
```
Comment (eladnava) [inline themes/neopolitan/index.html]: Another visual bug, too much padding in neopolitan:

screen shot 2016-10-01 at 11 38 14 pm

The two `<br>` statements should be moved to line 303 -- outside the `<% if (locals.action) { %>` closure.

**Comment by collnwalkr** [comment]
Good catch. Those changes have been made!

**Comment by eladnava** [comment]
@collnwalkr Thank you so much for your hard work and dedication on this PR! This was definitely not an easy PR, thanks for sticking with it and bearing with my comments. 😄

**Comment by eladnava** [comment]
Published `v1.0.27` to npm 👍

**Comment by collnwalkr** [comment]
@eladnava no worries! I really appreciate it when people are particular with code style. It helps me grow and learn. This is a cool project you have going on here. Glad I had the opportunity to help!

**Comment by eladnava** [comment]
Thanks so much for your kind words! 😄  Let me know if you'd like to work on anything else, we currently have 2 more outstanding issues that could use your crafty hands! 👍

**Comment by eladnava** [comment]
Published this feature in v2.0.0.
